src/services/formatters/YScopeFormatter/index.ts

```diff
@@ -0,0 +1,109 @@
+import {
+    BACKSLASH_REGEX,
+    FIELD_PLACEHOLDER_REGEX,
+    Formatter,
+    FormatterOptionsType,
+    YScopeFieldFormatter,
+    YScopeFieldPlaceholder,
+} from "../../../typings/formatters";
+import {LogEvent} from "../../../typings/logs";
+import {
+    getFormattedField,
+    splitFieldPlaceholder,
+    YSCOPE_FORMATTERS_MAP,
+} from "./utils";
+
+
+/**
+ * A formatter that uses a YScope format string to format log events into a string. See
+ * `YScopeFormatterOptionsType` for details about the format string.
+ */
+class YScopeFormatter implements Formatter {
+    #formatString: string;
+
+    #fieldPlaceholders: YScopeFieldPlaceholder[] = [];
+
+    constructor (options: FormatterOptionsType) {
+        this.#formatString = options.formatString;
+        this.#parseFieldPlaceholder();
```

Comment:
To avoid recreating the regexes on every call of `formatLogEvent`, can we use this pattern in `#parseFieldPlaceholder`?
```
const placeHolderPattern = new RegExp(FIELD_PLACEHOLDER_REGEX, "g");
const it = this.#formatString.matchAll(placeHolderPattern);
for (const match of it) {
    // The 1-index of exec result is the capture group in `FIELD_PLACEHOLDER_REGEX`.
    // (i.e. entire field-placeholder excluding braces).
    const [, placeholderString]: (string | undefined) [] = match;

    if ("undefined" === typeof placeholderString) {
        throw Error("Field placeholder regex is invalid and does not have a capture group");
    }

    const {fieldNameKeys, formatterName, formatterOptions} =
        splitFieldPlaceholder(placeholderString);

    let fieldFormatter: Nullable<YScopeFieldFormatter> = null;
    if (null !== formatterName) {
        const FieldFormatterConstructor = YSCOPE_FORMATTERS_MAP[formatterName];
        if ("undefined" === typeof FieldFormatterConstructor) {
            throw Error(`Formatter ${formatterName} is not currently supported`);
        }
        fieldFormatter = new FieldFormatterConstructor(formatterOptions);
    }

    this.#fieldPlaceholders.push({
        fieldNameKeys: fieldNameKeys,
        fieldFormatter: fieldFormatter,
        position: {
            startIndex: match.index,
            endIndex: match.index + placeholderString.length
        },
    });
}

```

Reply:
Then in `formatLogEvent()` we can use the `startIndex` and `endIndex` directly.